Documentation about upgradable contracts (#154)
Supercolony-net/openbrush-contracts: docs/docs/smart-contracts/upgradeable.md

---
sidebar_position: 1
title: Upgradeable contract
---

# Upgradeable contract

## Overview

Smart contracts are immutable by default, adding a layer of security and trust to the 
contracts. But software quality depends on the ability to upgrade source code to 
produce iterative releases. A certain degree of mutability is needed for bug fixing 
and potential product improvements.

> Upgradeability allows for experimenting and deploying the product at the early stage, 

Reviewer: ```suggestion
Upgradeability allows experimenting and deploying the product at the early stage, 
```

always leaving the chance to fix vulnerabilities and progressively add features. 
It is more actual right now while ink! and contract-pallet under active development. 

> Upgradeable contracts are not a Bug if they are developed consciously with 

Reviewer: ```suggestion
Upgradeable contracts are not a bug if they are developed consciously with 
```

Author: I tried to highlight it with capital letter=)

Decentralization can be achieved by providing the right to upgrade only to 
decentralized authority like governance, multisig, or another analog.

It is not hard to upgrade the logic of contracts. It can be achieved via `Proxy` and `Diamond` 
patterns or via the `set_code_hash` function provided by contract-pallet. The hardest 
part is to save the contract's state and make it compatible with new logic.

## Storage layout

### How storage works

Contracts use key-value storage to persist data. Each field of the contract may have 
its key and occupy storage cell. It is called storage layout.

![](assets/20220719_075309_F016550A-65D2-4DCD-A60A-D1A70B38D813.jpeg)

During compilation ink! inserts code to work with storage and ink! knows how to store 
each type in which storage cell. How exactly it works is not a part of that tutorial. 
The main point is that each type knows how to operate with each field and operate with 
storage, cause of a unique identifier. In the old version of ink! the identifier is 
`[u8; 32]` in a [new version](https://github.com/paritytech/ink/issues/1134) it is `u32`.

So, each data is stored under its unique identifier - the storage key. The value of the 

> key is the sequence of bytes - serialized(by SCALE codec) data type. The logic layer 

Reviewer: ```suggestion
key is the sequence of bytes - serialized (by SCALE codec) data type. The logic layer 
```

knows how to serialize and deserialize each data type. So during the execution, the 
logic layer deserializes all data by their storage keys and returns the 
filled contract's storage variable. The developer works with that variable, and before 
the end of the execution, the logic layer serializes data into sequences of bytes and 
stores them in their storage cells.

### Rules for upgradeable storage layout

The contract can have several logic layers (like in the `Diamond` pattern). 
So we will define rules in terms of several layers, but it is also applicable 
for upgradeable contracts with the `Proxy` pattern or `set_code_hash`.

1. The data stored under the storage key should use the same serialization and 
deserialization mechanism across all logic layers. Otherwise, some logic layers will not 
be able to deserialize the data type and fail.

> 1. Each logic unit(in most cases, it is a type) should occupy the same storage key across

Reviewer: ```suggestion
1. Each logic unit (in most cases, it is a type) should occupy the same storage key across
```

Following those simple rules will save you from corrupting the storage. 
Those rules are applicable for upgraded logic layers too.

If you use the same storage layout across all logic layers, and you don't plan to have 
unique fields per layer(so you don't plan to modify the storage layout in future upgrades). 
Then you already follow those rules with automatically calculated storage keys. 
But if you want to use a unique layout per layer or 
plan to modify the layout in the future, the next section will help you.

### Suggestions on how follow the rules

#### Approach description

The manual setting of the storage key for each field allows following the rules but 
makes development harder. ink! allows you to manually implement all utility traits and 
specify the storage key you want to use for each field. If your contract has 20 fields, 
then you need to set 20 storage keys.

The main suggestion is to design your storage layout as a bunch of logic units and 
assign a unique storage key per logic unit. The logic unit can be one field or a bunch 
of fields. In the scope of the logic unit, you can use automatically calculated keys 

> offsetted with the storage key of the logic unit, or you can use the same approach 

Reviewer 2: ```suggestion
offset with the storage key of the logic unit, or you can use the same approach 
```

#### Delegate call

Those contracts use [`delegate_call`](https://github.com/paritytech/substrate/issues/10566) to forward calls to logic layers. 
The delegate call accepts the logic layer's code hash and executes the source code, corresponding to the hash, 
in the context of the current contract. The source code works with the current contract state in that case. 
You can't call the constructor during a delegate call, so you can't naturally initialize the contract.

#### Workaround

Instead of using the constructors as the primary way to initialize the logic 
units(each logic layer has its logic unit or a bunch of logic units), 
you can add a separate initializer method and leave the constructor empty(with 
resolving [issue](https://github.com/paritytech/ink/issues/1187) you can not 
have constructor at all). That initialize method can accept any arguments the 
same as a typical constructor.

For example, for an upgradeable `PSP22` contract, you can add the `init_with_supply` method:

```rust
#[ink(message)]
pub fn init_with_supply(&mut self, total_supply: Balance) -> Result<(), PSP22Error> {
    self._mint(Self::env().caller(), total_supply)
}
```

> You can add as many methods as you wish with any arguments(the same as constructors).

Reviewer: ```suggestion
You can add as many methods as you wish with any arguments (the same as constructors).
```

contract, but you have only one facet. You can read about it in [that section](#logic-units-for-facet). 

### Usage of `set_code_hash` method

ink! has the `ink_env::set_code_hash` method, which allows replacing the code hash 
of the current contract. So you can change the logic layer by specifying a new code 
hash on demand. After setting a new code hash, the next call to your contract will execute updated logic.

You only need to expose that method somehow, and your common contract is upgradeable.
For example, you can add that method, and it is done:

```rust
#[ink(message)]
pub fn upgrade_my_contract(&mut self, new_code_hash: Hash) {
    ink_env::set_code_hash(&new_code_hash)
}
```

You need to consider the permission system because the only restricted set of people 
should be able to call that function.

All suggestions described above are applicable for that kind of upgradeable contracts. 
Better to have an upgradeable storage layout, initialization function for new versions 
of your contract, permission system, etc.

### The `Diamond` Standard

Using `Diamond` Standard you can add support for several facets(logic layers) that 

> can be upgraded. [That standard](https://eips.ethereum.org/EIPS/eip-2535) came 

Reviewer 3: ```suggestion
can be upgraded. [This standard](https://eips.ethereum.org/EIPS/eip-2535) came 
```

from the ethereum network. It works in the same way in ink! but instead of the 
address of the logic layer, you need to use the code hash of the logic layer.

This is the illustration of the flow of the `Diamond` pattern:

![](assets/20220715_130335_47FD0F8D-60F3-4FDF-82F4-672402FDC5D1.jpeg)

Each method in the smart contract has a selector. It is used as an identifier during the 
smart contract call to execute the right logic.
Each facet has a list of selectors that describe which methods are supported by the facet. 
Each selector is unique and belongs only to one facet. So selectors of facets can't overlap. 
`Diamond` contract knows which facet is responsible for which selector and forwards each 

> call to the smart contract to the corresponding facet(logic layer).

Reviewer: ```suggestion
call to the smart contract to the corresponding facet (logic layer).
```

`Diamond` contract has a function `diamond_cut` that allows registering each facet.

OpenBrush provides default implementation for `Diamond` standard on ink!.
For more details you can check [Diamond](diamond.md).

All suggestions above ideally describe how to develop an upgradeable contract
with multi-logic layers and many logic units.

> So here we will describe how to write facets(logic layers) with OpenBrush.

Reviewer: ```suggestion
So here we will describe how to write facets (logic layers) with OpenBrush.
```

#### Logic units for facet

> Each facet(logic layer) can have zero, one, or many logic units that work 

Reviewer: ```suggestion
Each facet (logic layer) can have zero, one, or many logic units that work 
```

with storage.

> Each logic unit should have a unique storage key and be upgradeable(support initialization

Reviewer: ```suggestion
Each logic unit should have a unique storage key and be upgradeable (support initialization
```

on demand, use storage key as an offset for all inner fields). It can be a struct with 

> one or many fields(structs without fields are useless) or an enum with at least two 

Reviewer: ```suggestion
one or many fields (structs without fields are useless) or an enum with at least two 
```

> variants(an enum with one variant is a structure). You can define struct/enum with the 

Reviewer: ```suggestion
variants (an enum with one variant is a structure). You can define struct/enum with the 
```

`openbrush::upgradeable_storage`, it will fail during execution. Each field should be 
upgradeable in the facet.

As an example, we will define logic units for `PSP22` and `Ownable` facets.

```rust
// OpenBrush uses the same logic unit for the default implementation of the `PSP22` trait.
#[openbrush::upgradeable_storage(openbrush::storage_unique_key!(PSP22Data))]
pub struct PSP22Data {
    // Total supply of the `PSP22`
    pub supply: Balance,
    // Balance of each user
    pub balances: Mapping<AccountId, Balance>,
    // Allowance to send tokens from one user to another
    pub allowances: Mapping<(AccountId, AccountId), Balance>,
    // Reserved fields for future upgrades
    pub _reserved: Option<()>,
}

// OpenBrush uses the same logic unit for the default implementation of the `Ownable` trait.
// It simply stores the `AccountId` of the owner of the contract.
#[openbrush::upgradeable_storage(openbrush::storage_unique_key!(OwnableData))]
pub struct OwnableData {
    // Owner of the contract
    pub owner: AccountId,
    // Reserved fields for future upgrades
    pub _reserved: Option<()>,
}
```

> `PSP22Data` and `OwnableData` have their storage keys. Both contain additional fields, 
> unrelated to business logic, `_reserved` for future upgrades(it adds overhead in one byte).

Reviewer 3: ```suggestion
`PSP22Data` and `OwnableData` have their storage keys. Both contain an additional field 
unrelated to business logic `_reserved` for future upgrades(it adds overhead in one byte).
```

to combine your logic units in the contract as fields. Leave the constructor empty, 
add the initialization method and methods related to business logic.

Example uses logic units defined in the previous section.

```rust
#[openbrush::contract]
pub mod facet_a {
    ...

    #[ink(storage)]
    #[derive(SpreadAllocate)]
    pub struct FacetA {
        psp22: PSP22Data,
        ownable: OwnableData,
    }
    
    // Your own implementation of `PSP22` trait.
    impl PSP22 for FacetA {
        ...
        
        #[ink(message)]
        fn balance_of(&self, owner: AccountId) -> Balance {
            ...
        }
    }

    impl FacetA {
        #[ink(constructor)]
        pub fn new() -> Self {
            // Empty constructor, we do here nothing
            ink_lang::codegen::initialize_contract(|instance: &mut Self| {})
        }

        // Initialization method to grant `total_supply` tokens to someone.
        #[ink(message)]
        pub fn init_with_supply(&mut self, total_supply: Balance) {
            assert_eq!(Self::env().caller(), self.ownable.owner, "Only owner can init contract");
            ...
        }
    }
}
```

> You can deploy the code of the facet into the blockchain. After, you can 

Reviewer 3: ```suggestion
You can deploy the code of the facet to the blockchain. After that, you can 
```

register your facet in `Diamond` via the `diamond_cut` method or in `Proxy` via 
`change_delegate_call` method.

#### Interaction between facets

> During development, you can have cases when one logic layer(for example `FacetA`) needs to interact with 

Reviewer: ```suggestion
During development, you can have cases when one logic layer (for example `FacetA`) needs to interact with 
```

> another logic unit or logic layer(for example `FacetB`). 

Reviewer: ```suggestion
another logic unit or logic layer (for example `FacetB`). 
```

    pub struct FacetB {
        // You embed `PSP22Data` logic unit from `FacetA`. 
        // It works with the same storage as `FacetA`. 
        // So you have access to data of `FacetA`.
        psp22: PSP22Data,
        // Some data for `FacetB`.
        foo: BarData,
    }

    impl FacetB {
        #[ink(constructor)]
        pub fn new() -> Self {
            // Empty constructor, we do here nothing
            ink_lang::codegen::initialize_contract(|instance: &mut Self| {})
        }

        #[ink(message)]
        fn balance_of_owner_in_facet_a(&self, owner: AccountId) -> Balance {
            // It accesses to the balance of `owner` that is managed by `FacetA`.
            self.psp22.balances.get(&owner).unwrap_or_default()
        }
    }
}
```

Embedding of the logic unit grants access to its storage.

##### Remark about logic units and OpenBrush

All data structures for contracts provided by OpenBrush are upgradeable logic units. 
So contracts support upgradeability by default.

You can access the default implementation when you embed OpenBrush data structures 
into your contract. You can use default implementation internally, or you can make 
it public.

> For example, when you embed `psp22::Data` into your `Contract`. You can already 

Reviewer 2: ```suggestion
For example, when you embed `psp22::Data` into your `Contract`, you can already 
```
